.github/workflows/rust-test.yml

```diff
@@ -68,20 +72,66 @@ env:
   CARGO_HTTP_USER_AGENT: "shipyard ${{ secrets.CARGO_PRIVATE_REGISTRY_TOKEN }}"
 
 jobs:
+  build-matrix:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Build Test Matrix
+        id: build
+        working-directory: ${{ inputs.working-directory }}
+        run: |
+          if [ "${{ github.event_name }}" != "pull_request" ]; then
+            echo "matrix=[\".\"]" >> $GITHUB_OUTPUT
+            exit 0
+          fi
+
+          git fetch origin ${{ github.base_ref }}
+          git_diff=$(git diff --name-only -r origin/${{ github.base_ref }} HEAD)
+
+          if [ "${{ runner.debug }}" = "1" ]; then
+            git diff --name-only -r origin/${{ github.base_ref }} HEAD
+          fi
+
+          workspace_changed="";
+          workspace_changed+=$(git diff --name-only --relative origin/main HEAD -- src)
+          workspace_changed+=$(git diff --name-only --relative origin/main HEAD -- Cargo.toml)
+
+          if [ "$workspace_changed" != "" ]; then
+            echo "matrix=[\".\"]" >> $GITHUB_OUTPUT
+            exit 0
+          fi
+
+          changed_paths=($(git diff --name-only -r origin/${{ github.base_ref }} HEAD | \
+                           grep "^${{ inputs.crates-directory }}/" | cut -d / -f 2 | sort | uniq | xargs))
+          output_paths=''
+          for changed_path in "${changed_paths[@]}"; do
+            output_paths+=",\"${{ inputs.crates-directory }}/$changed_path\""
+          done
+
+          if [ output_paths = '' ]; then
+            output_paths=",\".\""
+          fi
+
+          echo "matrix=[${output_paths:1}]" >> $GITHUB_OUTPUT
+    outputs:
+      matrix: ${{ steps.build.outputs.matrix }}
```

Comment:
I think that this could benefits using https://github.com/hellofresh/action-changed-files

Reply:
I just finished the first raw draft of multi workspace repos :see_no_evil: 
Maybe switch to that action later when we implement advanced path triggers?